js/compiler.js

```diff
@@ -1,262 +1,285 @@
-                    else if (section_type.startsWith("HRAM"))
-                        total = 127;
-                    logFunction("Space left: " + section_type + "[" + bank_nr + "]: " + space + "  (" + (space / total * 100).toFixed(1) + "%)");
-                }
-            }
-            logFunction("Build done");
-            flushLog();
-            done_callback(rom_file, start_address, addr_to_line);
-        }
-    }
-})(globalThis.compiler);
+import * as storage from "./storage.js";
+
+import createRgbAsm from "../rgbds/rgbasm";
+import createRgbLink from "../rgbds/rgblink";
+import createRgbFix from "../rgbds/rgbfix";
+
+var busy = false;
+var repeat = false;
+var start_delay_timer;
+var done_callback;
+var log_callback;
+var error_list = [];
+var rom_symbols = [];
+var ram_symbols = [];
+var link_options = [];
+
+var line_nr_regex = /([\w\.]+)[\w\.\:~]*\(([0-9]+)\)/gi;
+
+var buffered_log = "";
+function logFunction(str) {
+	if (/^[^:\s]+:/.test(str)) {
+		// If the log starts with a label, we can emit the previously buffered one.
+		flushLog();
+		buffered_log = str.trim();
+	} else {
+		// If the log doesn’t start with a label, it’s a follow-up to the previous log,
+		// so add it to the buffer.
+		buffered_log += " " + str.trim();
+	}
+}
+
+function flushLog() {
+	unbufferedLogFunction(buffered_log);
+	buffered_log = "";
+}
+
+function unbufferedLogFunction(str) {
+	if (log_callback) log_callback(str);
+
+	if (
+		str.startsWith("error: ") ||
+		str.startsWith("ERROR: ") ||
+		str.startsWith("warning: ")
+	) {
+		var type = "error";
+		if (str.startsWith("warning: ")) type = "warning";
+
+		var line_nr_match = str.matchAll(line_nr_regex);
+		for (var m of line_nr_match) {
+			var error_line = parseInt(m[2]);
+			error_list.push([type, m[1], error_line, str]);
+		}
+	}
+}
+
+export function setLogCallback(callback) {
+	log_callback = callback;
+}
+
+export function compile(callback) {
+	done_callback = callback;
+	if (busy) {
+		repeat = true;
+	} else {
+		busy = true;
+		trigger();
+	}
+}
+
+export function getErrors() {
+	return error_list;
+}
+
+export function getRomSymbols() {
+	return rom_symbols;
+}
+
+export function getRamSymbols() {
+	return ram_symbols;
+}
+
+export function setLinkOptions(options) {
+	link_options = options;
+}
+
+function trigger() {
+	if (typeof start_delay_timer != "undefined") clearTimeout(start_delay_timer);
+	start_delay_timer = setTimeout(startCompile, 500);
+}
+
+function startCompile() {
+	if (log_callback) log_callback(null);
+	error_list = [];
+	rom_symbols = [];
+	ram_symbols = [];
+
+	var targets = [];
+	for (const name of Object.keys(storage.getFiles()))
+		if (name.endsWith(".asm")) targets.push(name);
+	runRgbAsm(targets, {});
+}
+
+function runRgbAsm(targets, obj_files) {
+	var target = targets.pop();
+	logFunction("Running: rgbasm " + target + " -o " + target + ".o -Wall");
+	createRgbAsm({
+		arguments: [target, "-o", "output.o", "-Wall", "-l"],
```

Comment:
Snuck `-l` in here to silence the warning we see on live.

Reply:
I would say the warnings being reported and hightlighted in the code are one of the features